### PR TITLE
apf: add additional latency into work estimate

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
add additional latency into `width` for a request 

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
Extra latency should not be part of width.
Before we make a big divergence between counting requests and counting seats in the metrics, we should finish adding all the seat-counting metrics.
It would be easier to impose the extra latency in apf_filter.go than inside the QueueSet.  The one advantage I see for imposing extra latency inside the QueueSet is the ability to have a divergence between the request-couting metrics and the seat-counting metrics.
A sleep in the request-handling goroutine will not get the desired behavior, which is to get a response to the client before the extra latency expires.
A PassiveClock does not have a sleep method.
Remember that PassiveClock is designed to facilitate testing, and QueueSet is tested with a FakeEventClock.
Remember that we have two clock packages.

**Comment by MikeSpreitzer** [comment]
https://github.com/kubernetes/kubernetes/issues/94738 - two clock packages with divergent history

**Comment by MikeSpreitzer** [comment]
To do a testable sleep we would promote EventClock into the clock package and add a Sleep method, which the fake one implements like the existing ClockWait function in queueset_test.go.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:85]: The first commit isn't controversial at all - let's just split it to the separate PR (well - let's just not add additionalLatency field to the Width type but leave everything else as is) and let's just merge it.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:804]: This isn't sleeping yet...

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:804]: OK - I see now that this is desired.
I think this requires a more extensive comments to describe what exactly is happening here.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:85]: @wojtek-t  sounds good, i just opened https://github.com/kubernetes/kubernetes/pull/103504

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:650]: Do we need the "executed" field?

I started wondering (didn't check that deeply enough yet) if we can't check that based on "decision" field being set to decisionExecute...

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:85]: I objected to the first commit in the first sentence of https://github.com/kubernetes/kubernetes/pull/103240#issuecomment-869802687

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:85]: But that commit doesn't add additional latency (i.e. the one that I approved didn't add it).

**Comment by wojtek-t** [comment]
OK - at this point of release, I no longer believe we can reliably have that in 1.22.
So let's postpone it until 1.23.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:85]: A struct named "Width" that only has one field seems a bit pointless.  (more extensive commentary elsewhere.)

**Review by MikeSpreitzer — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:829]: yes, I got stuck here and then started working on https://github.com/kubernetes/kubernetes/pull/103829.

**Comment by tkashem** [comment]
Next Steps:
- take `AdditionalLatency` in account for queuset dispatching logic
- wait for https://github.com/kubernetes/kubernetes/pull/103664 to merge and rebase
- wait for https://github.com/kubernetes/kubernetes/pull/103829 to use `EventClock` (https://github.com/kubernetes/kubernetes/pull/103830 needs to merge first), remove https://github.com/kubernetes/kubernetes/pull/103240/commits/af59872ea150b882b680c9fef5f57e5c45b5cb80 and rebase
-

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:835]: If possible I would like to avoid this goroutine but I don't see a way for it. Also - additionalLatency should be 0 for huge majority of the calls, so it's probably not that big issue.


That said - I would like to get @MikeSpreitzer comment if this goroutine will not break the goroutine counting that we currently have.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1022]: 100ms is definitely too short in starved prow environment. Please use waitTimeForever (or sth like that we have).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1028]: This is very error prone. I think my comment about goroutine counting above might be relevant here (i.e. wait for gorutine be finished given we already have this framework).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:835]: Indeed, the queueset code should participate in goroutine counting.  A relatively small fix would be to use the EventAfterDuration method of an EventClock, that way you get efficient code in the non-testing case and goroutine counting in test code.

It should be possible to eliminate the creation of a goroutine in the case where there are no requests queued.  In this case the release of seats can be delayed until something else prods the queueSet, at which time it can be decided whether the seats should have already been released or released in the future (again either via a goroutine or waiting for another prod).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:812]: This could be a method of `*queueSet`.  It _might_ be more efficient if we wrote it that way, depending on compiler smarts; it would be a miniscule difference at most.  I am OK with leaving it here for the sake of code locality.  I am OK with moving it out.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1028]: By using EventAfterDuration rather than Sleep, we can advance fake time as we please to exercise the code.

**Review by MikeSpreitzer — COMMENTED** [review]
i prefer behavioral tests over checksum tests.  How about a usage scenario that exercises the new functionality and measures the resultant behavior?

Other comments done independently.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: Yes, using `EventAfterDuration` now. But the test with additional latency is failing, I am investigating...

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go]: I don't think we have many requests with `AdditionalLatency` today, so the cost of doing it in a new goroutine may be negligible today. We can let this solution merge today, and I can do a separate PR in the future that uses a `prod queueSet` approach, thoughts? I will leave a TODO if this is acceptable.

**Comment by tkashem** [comment]
The unit test is failing, I am going to investigate it...

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:806]: EventAfterDuration spins up a goroutine itself - so you don't want the outer one.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1054]: I think you can just do clk.Run(nil)

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1054]: [and do this unconditionally to ensure that everything fired and finished]

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1059]: [and if you do above, I don't think you need locking]

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1060]: And following on @MikeSpreitzer comment, you can also check time now - it should be now+additional_latency at this point

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go]: good catch, that was causing the test to fail.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: I prefer `SetTime` because this is a way for the test to say - i expect the event to happen exactly after `AdditionalLatency`

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: yes, removed.

**Review by MikeSpreitzer — COMMENTED** [review]
`newTestableQueueSetFactory` is intended to be the start of how a test func creates a queueSet.
Note that when doing goroutine counting, you need to use a promiseFactoryFactory that participates.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:786]: How about calling this `releaseSeatsLocked` ?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1051]: Why not run clk.Run(nil)

[but then Mike`s comment about proper initialization of QS is a valid one]

**Review by wojtek-t — COMMENTED** [review]
The prod code looks fine to me.

The test still requires some though - see also Mike`s comments.

**Comment by MikeSpreitzer** [comment]
```
demands[i] = float64(nThreads) * float64(uc.execDuration) / float64(uc.thinkDuration+uc.execDuration)
```
would become
```
demands[i] = float64(nThreads) * float64(width) * float64(uc.execDuration) / float64(uc.thinkDuration + uc.padDuration + uc.execDuration)
```

**Comment by tkashem** [comment]
> Rather than a checksum test, I recommend updating the behavioral tests.

this test is a behavioral test, but at a unit level, so if there is a regression due to a bug in `finishRequestLocked` this test will alert us by failing. 
The behavioral test you are referring to is happening is exercising the code from a much higher level. So I think this test in this PR is essential.

I agree that we should update the behavioral tests with variable work estimates. Do we want to do it in a follow-up PR? I am trying to expedite this to unblock https://github.com/kubernetes/kubernetes/pull/103539. otherwise, happy to update the behavioral tests.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1051]: > `newTestableQueueSetFactory` is intended to be the start of how a test func creates a queueSet.
> Note that when doing goroutine counting, you need to use a promiseFactoryFactory that participates.


@MikeSpreitzer  do you mean this initialization pattern?
```
	clk, counter := testeventclock.NewFake(now, 0, nil)
	qsf := newTestableQueueSetFactory(clk, countingPromiseFactoryFactory(counter))
	qCfg := fq.QueuingConfig{
		Name:             "TestTotalRequestsExecutingWithPanic",
		DesiredNumQueues: 0,
		RequestWaitLimit: 15 * time.Second,
	}
	qsc, err := qsf.BeginConstruction(qCfg, newObserverPair(clk))
	if err != nil {
		t.Fatal(err)
	}
	qs := qsc.Complete(fq.DispatchingConfig{ConcurrencyLimit: 1})
	counter.Add(1) // account for the goroutine running this test
```

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1051]: > Why not run clk.Run(nil)

@wojtek-t for the test at hand, `clk.SetTime` is more concise in expressing the expectation that the event function should be executed immediately after `AdditionalLatency` elapses. So this is a verification step.

Alternatively, we can achieve this with `clk.Run(nil)` too:
```
clk.Run(nil)
get the time when the event fired and assert that it happened after AdditionalLatency elapsed
```

I think `SetTime` does the job concisely, thoughts?

**Comment by tkashem** [comment]
@MikeSpreitzer this test exercises `finishRequestLocked` (a unit at a much lower level), not `qs.StartRequest` - so i don't think we need to use the above pattern, it adds unnecessary steps for the test:
```
                        now := time.Now()
			clk, counter := testeventclock.NewFake(now, 0, nil)
			qsf := newTestableQueueSetFactory(clk, countingPromiseFactoryFactory(counter))
			qCfg := fq.QueuingConfig{
				DesiredNumQueues: 1,
				QueueLengthLimit: 8,
				HandSize:         1,
			}
			qsc, err := qsf.BeginConstruction(qCfg, newObserverPair(clk))
			if err != nil {
				t.Fatal(err)
			}
			o := qsc.Complete(fq.DispatchingConfig{ConcurrencyLimit: 1})
			counter.Add(1) // account for the goroutine running this test

			qs, ok := o.(*queueSet)
			if !ok {
				t.Fatal("Not a queueSet")
			}

			queue := qs.queues[0]
			r := &request{
				qs:           qs,
				queue:        queue,
				workEstimate: test.workEstimate,
			}

                        ...

			qs.finishRequestLocked(r)
			counter.Add(-1) // completion of main activity of goroutine running this test

```

I am fine making this change if we want to stick to creating a fully initialized QueueSet in every unit test (as opposed to initialize what is needed), thoughts?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1051]: In my opinion, using `SetTime` to the precise time needed is sufficient for a test of `finishRequestLocked`.  Using `clk.Run(nil)` would also be OK, and adding a test that `Run` advanced the fake time to the exact correct value would be better.

**Comment by MikeSpreitzer** [comment]
Yes, that's the initialization pattern I meant.

I suppose "checksum test" means "test of a small unit".  My point is that tests of the whole queueset unit are necessary and sufficient.  I am not a fan of maintaining tests of smaller units.  If you want to put in a test of `finishRequestLocked` with non-zero duration padding in this PR and switch to whole-package testing in a follow-on PR, that is OK with me.  I wiill prepare that follow-on.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1050]: If we really want to lock down the behavior of `finishRequestLocked`, we could also check the values one nanosecond before the proper release time.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:809]: I do not understand why local copy of this value is being created here.  As far as i can tell, it is harmless but not helpful.

**Review by MikeSpreitzer — APPROVED** [review]
I made some independent comments.
I do not want to hold up this improvement due to wrangling over the test; that can be worked in follow-up.

**Comment by MikeSpreitzer** [comment]
Thanks!

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:809]: `r` represents request scoped data and should not be accessed from another active goroutine, but just to be on the safe side i made a copy here.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1050]: Yes, I will update the test to assert on this.

**Comment by tkashem** [comment]
> I do not want to hold up this improvement due to wrangling over the test; that can be worked in follow-up.

Thanks @MikeSpreitzer, I will be working on the follow-on PR.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:809]: Neither the original nor the copy is accessed from the forked goroutine.

**Pull request merged**
